src/paths/voronoi.c

```diff
@@ -0,0 +1,423 @@
+                    IGRAPH_CHECK(igraph_2wheap_push_with_index(&q, to, -altdist));
+                } else if (igraph_2wheap_has_active(&q, to)) {
+                    igraph_real_t curdist = -igraph_2wheap_get(&q, to);
+                    if (altdist < curdist) {
+                        /* This is a shorter path */
+                        igraph_2wheap_modify(&q, to, -altdist);
+                    }
+                }
+            }
+        } /* !igraph_2wheap_empty(&q) */
+    }
+
+    if (tiebreaker == IGRAPH_VORONOI_RANDOM) {
+        igraph_vector_int_destroy(&tie_count);
+        IGRAPH_FINALLY_CLEAN(1);
+    }
+
+    igraph_2wheap_destroy(&q);
+    igraph_inclist_destroy(&il);
+    IGRAPH_FINALLY_CLEAN(2);
+
+    return IGRAPH_SUCCESS;
+}
+
+
+/**
+ * \function igraph_voronoi
+ * \brief Voronoi partitioning of a graph.
+ *
+ * \experimental
+ *
+ * To obtain a Voronoi partitioning of a graph, we start with a set of generator
+ * vertices, which will define the partitions. Each vertex is assigned to the generator
+ * point from (or to) which it is closest.
+ *
+ * </para><para>
+ * This function uses a BFS search for unweighted graphs and Dijkstra's algorithm
+ * for weights ones.
+ *
+ * \param graph The graph to partition.
+ * \param membership If not \c NULL, the Voronoi partition of each vertex
+ *    will be stored here. <code>membership[v]</code> will be set to the index
+ *    in \p generators of the generator point that \c v belongs to. For vertices
+ *    that are not reachable from any generator, <code>-1</code> is returned.
+ * \param distances If not \c NULL, the distance of each vertex to its respective
+ *    generator will be stored here. For vertices which are not reachable from
+ *    any generator, \c IGRAPH_INFINITY is returned.
+ * \param generators Vertex IDs of the generator points.
+ * \param weights The edge weights, interpreted as lengths in the shortest
+ *    path calculation. All weights must be non-negative.
+ * \param mode In directed graphs, whether to compute distances \em from
+ *    generator points to other vertices (\c IGRAPH_OUT), \em to generator
+ *    points from other vertices (\c IGRAPH_IN), or ignore edge directions
+ *    entirely (\c IGRAPH_ALL).
+ * \param tiebreaker Controls which generator point to assign a vertex to
+ *    when it is at equal distance from/to multiple generator points.
+ *    \clist
+ *    \cli IGRAPH_VORONOI_FIRST assign the vertex to the first generator point.
+ *    \cli IGRAPH_VORONOI_LAST assign the vertex to the last generator point.
+ *    \cli IGRAPH_VORONOI_RANDOM assign the vertex to a random generator point.
+ *    \endclist
+ *    Note that \c IGRAPH_VORONOI_RANDOM does not guarantee that all vertices
+ *    of the subgraphs induced by the partitions will be reachable from the
+ *    generator. For example, if 1 and 2 are chosen as genrators for the graph
+ *    <code>4-3-2, 2-1</code>, then 3 and 4 may be randomly assigned to different
```

Comment:
I'm not sure I understand this line, should it be
`<code>4-3, 2-1</code>`
maybe?

Reply:
That was meant to be shorthand for `4-3, 3-2`, but the example is wrong. I will fix it, thanks for noticing.

It should be `1-3, 2-3, 3-4`, for this graph:

<img width="127" alt="image" src="https://user-images.githubusercontent.com/1212871/193410535-f356b9f3-3a6d-4d3f-b3e6-d5d0a01fa34f.png">

If the generators are 1 and 2, then 3 and 4 are at equal distance from both of them. The random assignment may cause 3 to be assigned to 2 and 1 to 4. Like this:

<img width="127" alt="image" src="https://user-images.githubusercontent.com/1212871/193410570-40f00950-a4f4-41c6-a1ac-5ef15cda3db9.png">

That means that the partition that belongs to 1 is no longer contiguous. If we take the corresponding subgraph, it will not be connected. This is "not nice", and I though about possible solutions, but it seems far from trivial, so I decided to leave it for later.

Reply:
I tried to clarify this paragraph, hopefully it's better now.